### DISCUSSION TITLE
Proposal for a "control" domain and possible generic integrations.

### DISCUSSION THREAD

**Discussion by wonderslug** [home-assistant/architecture#926] [General]
Howdy all,

I would like to suggest the addition of a **control** domain.  A **control** domain would allow for the implementation of generic control patterns as well as provide a generic home for entities that don't fit into the current entity model.

### Generic Control Patterns
---
A lot of automations that get created are really just very common control patterns. A lot of times we implement these patterns very poorly, at least I know I do.  To have some of these patterns as a pre-defined control with feedback loops built in would allow for some quick implementations of some relatively complex systems very quickly.  We currently have a lot of systems around use that use these sorts of model, humidifier and thermostats are ones that have dedicated controls within HA, but things like plant watering based on soil moisture levels, battery charging based on voltage level or other cusotm developed end user system can benefit from the same type of patterns.

As an example, one of my current use-cases is a mechanism to the free chlorine level in my pool within a range.  I have a Salt Water Generator that generates chlorine for the pool by running and converting the salt in the salt water pool free chlorine.  But it is timed and does not have a feedback mechanism to understand what the chlorine level actually is.  Now I can get the free chlorine level every 8 hours using the Sutro integration (HACS) and be able to use that to respond and control the Salt Water Generator.  This becomes a system very similar to a thermostat for climate control but with different measurement units and time ranges for reaction.  When the level gets too low and outside a certain amount of tolerance (like with a heater mode on a thermostat) it kicks on the generator to start adding more.

Now this could be handled with scripts, automations, or blueprints, but so could a thermostat.  A **control** entity would allow for the programmatic encapsulation of the a different types of well known and commonly used control behaviours and patterns within an integration and controllable entity. In my mind, the feedback mechanisms with thresholds and tolerance levels really drive this sort of functionaility to be packaged into a simple implementation

Some possible generic integrations would be things like:

- negative-feedback-control
   This is essentially equivalent to a simple thermostat (generic-thermostat), with a threshold value, tolerance, and entity trigger to keep a sensor above/below the threshold.

- ranged-negative-feedback-control
  This would be like a auto ranged thermostat kicking on high and low tigger entities to maintain the sensor in range.

- binary-timed-control
  This would follow a binary sensor and trigger and entity until the binary sensor is flipped back or a time limit has been reached.  And example of this is an auto fill leveler that would trigger when the level gets to low and we want to maintain kick something on to fill to the right level but make sure that it doest use to much to fill in one time period. (Im focused on my pool automation right now.  Can ya tell?)

As well there are other possible implementations based on various hysteresis models or feedback control mechanisms since encapsulation in a control entity would allow for historical data to be processed and used along with maintenance of state for things like averages or extremes without the need to implement and store in helpers.

### Generic Landing Spot
---
Having a **control** domain would allow a place for a lot of non-bucketed entities to land.  Entities that don't really fit in the defined set of domains don't really have category that they can be reside in.  A generic **control** domain would allow for the generic patterns talked about above, but also allow for the "orphans" of taxonomy a place to be implemented in.

### Commonality
---
There seems to be only a simple set of common services that the **control** domain could implement that could apply to all entities implemented, such as maybe enable and disable, but the pattern has seemed to be that dedicated integration services have been gaining a lot more traction.  This would probably meet the needs of the generic integrations and allow the flexibility to implement and innovate to other implemented **control** entities and integrations.

**Comment by parautenbach** [comment]
Interesting idea, especially the negative feedback control, which is a classical control system mechanism. I'll keep following the conversation. 

> ranged-negative-feedback-control
> This would be like a auto ranged thermostat kicking on high and low tigger entities to maintain the sensor in range.

I just wanted to point out that I think the [threshold integration](https://www.home-assistant.io/integrations/threshold/) fulfils this (at least in part).

**Comment by wonderslug** [comment]
@parautenbach 

Sorry, my bad with deleted one... I didnt read correctly.

Yeah the threshold integration gives you the place to get a sensor out of it and trigger an automation but personally I would really like to see this as an entity that I can configure with like [generic_thermostat](https://www.home-assistant.io/integrations/generic_thermostat/) but with the ability to do other measurements and threshold types.
